packages/apps/web-viewer-test/src/providers/EcsqlWidget.tsx

```diff
@@ -0,0 +1,147 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { useActiveIModelConnection } from "@itwin/appui-react";
+import { QueryRowFormat } from "@itwin/core-common";
+import type { IModelConnection } from "@itwin/core-frontend";
+import {
+  Button,
+  ExpandableBlock,
+  LabeledTextarea,
+  Table,
+} from "@itwin/itwinui-react";
+import React, { useState } from "react";
+
+export default function EcsqlWidget() {
+  const [input, setInput] = useState("");
+  const [data, setData] = useState<Record<string, string | React.ReactNode>[]>(
+    []
+  );
+  const [headers, setHeaders] = useState<string[]>([]);
+  const [isLoading, setIsLoading] = useState(false);
+  const [error, setError] = useState("");
+  const iModelConnection = useActiveIModelConnection() as IModelConnection;
+
+  const getData = async () => {
+    try {
+      setData([]);
+      setHeaders([]);
+      setIsLoading(true);
+      const results = iModelConnection.query(input, undefined, {
+        rowFormat: QueryRowFormat.UseJsPropertyNames,
+      });
+      const rows: Record<string, string | React.ReactNode>[] = [];
+      const columnHeaders: string[] = [];
+
+      for await (const obj of results) {
+        const row = { ...obj };
+        // Case 1: Need to convert booleans to strings to display them in the table
+        Object.keys(row)
+          .filter((key) => typeof row[key] === "boolean")
+          .forEach((key) => {
+            row[key] = obj[key].toString();
+          });
+        // Case 2: Need to stringify objects to display them in the table
+        Object.keys(row)
+          .filter((key) => typeof row[key] === "object")
+          .forEach((key) => {
+            row[key] = JSON.stringify(obj[key]);
+          });
+        // Case 3: If the cell content is too long, put it inside an ExpandableBlock
+        Object.keys(row)
+          .filter(
+            (key) => typeof row[key] === "string" && row[key].length > 500
+          )
+          .forEach((key) => {
+            row[key] = (
+              <ExpandableBlock title="..." size="small" styleType="borderless">
+                {row[key]}
+              </ExpandableBlock>
+            );
+          });
+
+        rows.push(row);
+        for (const key of Object.keys(row)) {
+          if (!columnHeaders.includes(key)) {
```

Comment:
well wouldnt they always all not be in columnheaders to begin with?
re-read this block of code here

Reply:
Hmm, I'm pretty sure this is right. I wanted columnHeaders to be the superset of all keys in each row, because sometimes each row might have different keys. So every time columnHeaders encounters a new unique key, it should append the new key. columnHeaders is instantiated before we start iterating through each row.